src/tqec/sketchup/block_graph.py

```diff
@@ -0,0 +1,551 @@
+
+    def normal_direction_to_corner_plane(self) -> Direction3D:
+        """If the cube is at a corner, return the normal direction to the corner plane.
+
+        Due to the color match rule at the corner turn, the cornel plane can be inferred
+        from the type of the cube.
+        """
+        if self == CubeType.VIRTUAL:
+            raise TQECException("Cannot infer the corner plane for a virtual cube.")
+        if self.value.count("z") == 2:
+            return Direction3D.from_axis_index(self.value.index("x"))
+        return Direction3D.from_axis_index(self.value.index("z"))
+
+    def infer_pipe_type_at_direction(
+        self,
+        direction: Direction3D,
+        src_side: bool = True,
+        has_hadamard: bool = False,
+    ) -> PipeType:
+        """Infer the pipe type connecting this cube at some direction with the color match rule."""
+        if self == CubeType.VIRTUAL:
+            raise TQECException("Cannot infer the pipe type for a virtual cube.")
+        color = self.get_color().pop_color_at_direction(direction)
+        return PipeType.from_color_at_side(color, src_side, has_hadamard)
+
+
+class PipeType(Enum):
+    """Valid pipe types in the library."""
+
+    # pipes without H
+    OZX = "ozx"
+    OXZ = "oxz"
+    XOZ = "xoz"
+    ZOX = "zox"
+    XZO = "xzo"
+    ZXO = "zxo"
+    # pipes with H
+    OZXH = "ozxh"
+    OXZH = "oxzh"
+    ZOXH = "zoxh"
+    XOZH = "xozh"
+    ZXOH = "zxoh"
+    XZOH = "xzoh"
+
+    @property
+    def has_hadamard(self) -> bool:
+        """Check if the block type has a hadamard transition."""
+        return "h" in self.value
+
+    @property
+    def direction(self) -> Direction3D:
+        """Return the direction of the pipe."""
+        return [Direction3D.X, Direction3D.Y, Direction3D.Z][self.value.index("o")]
+
+    def get_color_at_side(self, src_side: bool = True) -> Color3D:
+        """Get the color of the pipe at the given side."""
+        if not self.has_hadamard or src_side:
+            return Color3D.from_string(self.value[:3])
+        return Color3D.from_string(self.value[:3], flip_xz=True)
+
+    @staticmethod
+    def from_color_at_side(
+        color: Color3D, src_side: bool = True, has_hadamard: bool = False
+    ) -> "PipeType":
+        """Get the pipe type from the color at one side."""
+        if not sum(c is None for c in astuple(color)) == 1:
+            raise TQECException(
+                "Exactly one color must be undefined for a pipe along the pipe direction."
+            )
+        pipe_color = []
+        for c in astuple(color):
+            if c is None:
+                pipe_color.append("o")
+            elif has_hadamard and not src_side:
+                pipe_color.append("x" if c == "z" else "z")
+            else:
+                pipe_color.append(c)
+        if has_hadamard:
+            pipe_color.append("h")
+        return ty.cast(PipeType, parse_block_type_from_str("".join(pipe_color)))
```

Comment:
```suggestion
        return PipeType("".join(pipe_color).lower())
```
is enough right? The `parse_block_type_from_str` seems to be adding more code than it removes.